File: minimum-genetic-mutation/minimum-genetic-mutation.cs
public class Solution {
    
    public bool singleDiffChar(string gene1, string gene2)
    {
        int cnt=0;
        for(int i=0;i<gene1.Length;i++)
        {
            if(gene1[i]!=gene2[i])
                cnt++;
        }
        
        return cnt==1;
    }
    
    public int MinMutation(string startGene, string endGene, string[] bank) {
        
        Dictionary<string,List<string>> map=new Dictionary<string, List<string>>();
        
        if(bank.Length==0)
            return -1;
        
        for(int i=0;i<bank.Length;i++)
        {
            for(int j=i+1;j<bank.Length;j++)
            {
                if(singleDiffChar(bank[i],bank[j]))
                {
                    if(!map.ContainsKey(bank[i]))     
                        map.Add(bank[i],new List<string>());
                    
                    if(!map.ContainsKey(bank[j]))
                        map.Add(bank[j],new List<string>());
                    
                    map[bank[i]].Add(bank[j]);
                    map[bank[j]].Add(bank[i]);
                }
            }
        }
        
        for(int i=0;i<bank.Length;i++)
        {
            if(singleDiffChar(bank[i],startGene))
            {
                if(!map.ContainsKey(bank[i]))     
                        map.Add(bank[i],new List<string>());
                    
                if(!map.ContainsKey(startGene))
                        map.Add(startGene,new List<string>());
                    
                    map[bank[i]].Add(startGene);
                    map[startGene].Add(bank[i]);
            }
        }
        
        
        HashSet<string> seenSet=new HashSet<string>();
        
        Queue<string> bfs=new Queue<string>();
        
        bfs.Enqueue(startGene);
        int level=0;
        
        while(bfs.Count()>0)
        {
            int len=bfs.Count();
            level++;    
            for(int i=0;i<len;i++)
            {
                string currGene=bfs.Dequeue();
                seenSet.Add(currGene);
                foreach(var adjGene in map[currGene])
                {
                    if(!seenSet.Contains(adjGene))
                        bfs.Enqueue(adjGene);
                    seenSet.Add(adjGene);
                }
            }
            
            if(seenSet.Contains(endGene))
                return level;
        }
        
        return -1;
        
    }
}

File: minimum-genetic-mutation/minimum-genetic-mutation.java
class Mutation
{
    String currGene;
    int steps;
    
    public Mutation(String Gene,int step)
    {
        currGene=Gene;
        steps=step;
    }
}

class Solution {
    
     HashSet<String> seen=new HashSet<>();
    
    public boolean getDiff(String str1,String str2)
    {
        int diffChar=0;
        for(int i=0;i<str1.length();i++)
        {
            if(str1.charAt(i)!=str2.charAt(i))
            {
                diffChar++;
            }
        }
        
        return diffChar==1;
    }
    
    public int minMutation(String startGene, String endGene, String[] bank) {
        
        HashMap<String, List<String>> map=new HashMap<>();
        
        if(bank.length==0)
            return -1;
        
        for(int i=0;i<bank.length;i++)
        {
            for(int j=0;j<bank.length;j++)
            {
                if(getDiff(bank[i],bank[j]))
                {
                    if(!map.containsKey(bank[i]))
                    {
                        map.put(bank[i],new ArrayList<String>());
                    }
                    if(!map.containsKey(bank[j]))
                    {
                        map.put(bank[j],new ArrayList<String>());
                    }
                    
                    map.get(bank[i]).add(bank[j]);
                    map.get(bank[j]).add(bank[i]);
                    
                }
            }
        }
        
        for(int i=0;i<bank.length;i++)
        {
            if(getDiff(bank[i],startGene))
            {
                if(!map.containsKey(startGene))
                {
                    map.put(startGene,new ArrayList<String>());
                }
                
                if(!map.containsKey(bank[i]))
                {
                    map.put(bank[i],new ArrayList<String>());
                }
                
                map.get(startGene).add(bank[i]);
                map.get(bank[i]).add(startGene);
            }
        }
        
       
        //seen.add(startGene);
        //dfs(startGene, endGene, map);
        
        Stack<Mutation> stack=new Stack<>();
        
        stack.add(new Mutation(startGene,0));
        
        while(stack.size()>0)
        {
            Mutation curr=stack.pop();
            //System.out.println(curr.currGene);
            for(String gene: map.get(curr.currGene))
            {
                if(!seen.contains(gene))
                {
                    seen.add(gene);
                    stack.push(new Mutation(gene,curr.steps+1));
                   // System.out.println(curr.currGene+"-"+gene);

                    if(gene.equals(endGene))
                        return curr.steps+1;
                }
            }
        }
        
        return -1;
    }
    
}
